server.go

```diff
@@ -800,7 +800,7 @@ func (s *Server) processUnaryRPC(t transport.ServerTransport, stream *transport.
 	}
 	opts := &transport.Options{
 		Last:  true,
-		Delay: false,
+		Delay: true,
```

Comment:
I agree. It seems like the calls to `adjustNumWriter` and the associated flushing are at too low a level. For unary RPCs, the number of writers should be incremented before calling `sendResponse` and decremented after the `WriteStatus` calls. The flushing logic currently in `sendResponse` is a bit difficult to reason about given the interaction with flow control and it isn't obvious to me how to pull it to a higher level. Perhaps this is the subtle bug you're referring to. 

I'm happy to leave this work in your hands. Consider this PR more an indication of the importance of this work. The unnecessary flushes introduce a significant amount of latency!